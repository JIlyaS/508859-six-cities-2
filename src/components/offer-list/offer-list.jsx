import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import ActionCreator from '../../action-creator/action-creator';
import OfferCard from '../offer-card/offer-card';
import Operation from '../../operation/operation';
import {getClassOfferCardName, getLocalStorageLogin} from '../../utils';

class OfferList extends PureComponent {
  constructor(props) {
    super(props);

    this._changeFavoriteOfferClickHandler = this._changeFavoriteOfferClickHandler.bind(this);
  }

  render() {
    const {offers, classOfferCard, activeOfferMouseEnterHandler, deactiveOfferMouseLeaveHandler} = this.props;
    const classCard = getClassOfferCardName(classOfferCard);
    return <div className={classCard.list}>
      {offers.map((offer) => <OfferCard
        offerId={offer.id}
        offer={offer}
        key={offer.id}
        activeOfferMouseEnterHandler={activeOfferMouseEnterHandler}
        deactiveOfferMouseLeaveHandler={deactiveOfferMouseLeaveHandler}
        changeFavoriteOfferClickHandler={this._changeFavoriteOfferClickHandler}
        classCard={classCard}
      />)}
    </div>;
  }

  _changeFavoriteOfferClickHandler(offerId, isFavorite) {
    const {changeOfferFavorite, getLogin} = this.props;
    getLogin();
    const login = getLocalStorageLogin();
    if (login) {
      const status = isFavorite === true ? 0 : 1;
      changeOfferFavorite(offerId, status);
    }
  }
}

OfferList.propTypes = {
  offers: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  activeOfferMouseEnterHandler: PropTypes.func.isRequired,
  deactiveOfferMouseLeaveHandler: PropTypes.func.isRequired,
  changeOfferFavorite: PropTypes.func.isRequired,
  getLogin: PropTypes.func.isRequired,
  classOfferCard: PropTypes.string.isRequired,
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  activeOfferCard: state.userReducer.activeOfferCard,
});

const mapDispatchToProps = (dispatch) => ({
  activeOfferMouseEnterHandler: (offerCard) => dispatch(ActionCreator.changeActiveCard(offerCard)),
  deactiveOfferMouseLeaveHandler: () => dispatch(ActionCreator.removeActiveCard()),
  changeOfferFavorite: (offerId, status) => dispatch(Operation.changeOfferFavorite(offerId, status)),
  getLogin: () => dispatch(Operation.getLogin()),
});

export {OfferList};

export default connect(mapStateToProps, mapDispatchToProps)(OfferList);
