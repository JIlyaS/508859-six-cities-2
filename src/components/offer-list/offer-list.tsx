import * as React from 'react';
import {connect} from 'react-redux';

import ActionCreator from '../../action-creator/action-creator';
import OfferCard from '../offer-card/offer-card';
import Operation from '../../operation/operation';
import {getClassOfferCardName, getLocalStorageLogin} from '../../utils';
import {OfferListProps} from '../../types/types';

class OfferList extends React.PureComponent<OfferListProps, null> {
  constructor(props) {
    super(props);

    this.handleFavoriteOfferClick = this.handleFavoriteOfferClick.bind(this);
  }

  handleFavoriteOfferClick(offerId, isFavorite) {
    const {onChangeOfferFavorite, onGetLogin} = this.props;
    onGetLogin();
    const login = getLocalStorageLogin();
    if (login) {
      const status = isFavorite === true ? 0 : 1;
      onChangeOfferFavorite(offerId, status);
    }
  }

  render() {
    const {offers, classOfferCard, onActiveOfferMouseEnter, onDeactiveOfferMouseLeave} = this.props;
    const classCard = getClassOfferCardName(classOfferCard);
    return <div className={classCard.list}>
      {offers.map((offer) => <OfferCard
        offerId={offer.id}
        offer={offer}
        key={offer.id}
        onActiveOfferMouseEnter={onActiveOfferMouseEnter}
        onDeactiveOfferMouseLeave={onDeactiveOfferMouseLeave}
        onFavoriteOfferClick={this.handleFavoriteOfferClick}
        classCard={classCard}
      />)}
    </div>;
  }
}

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  activeOfferCard: state.userReducer.activeOfferCard,
});

const mapDispatchToProps = (dispatch) => ({
  onActiveOfferMouseEnter: (offerCard) => dispatch(ActionCreator.changeActiveCard(offerCard)),
  onDeactiveOfferMouseLeave: () => dispatch(ActionCreator.removeActiveCard()),
  onChangeOfferFavorite: (offerId, status) => dispatch(Operation.changeOfferFavorite(offerId, status)),
  onGetLogin: () => dispatch(Operation.getLogin()),
});

export {OfferList};

export default connect(mapStateToProps, mapDispatchToProps)(OfferList);
