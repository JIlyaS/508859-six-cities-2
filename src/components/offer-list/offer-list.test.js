import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import {shape} from 'prop-types';
import {BrowserRouter} from 'react-router-dom';

import {OfferList} from './offer-list';

Enzyme.configure({adapter: new Adapter()});

const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
};

const createContext = () => ({
  context: {router},
  childContextTypes: {router: shape({})},
});

it(`OfferList correctly renders after relaunch`, () => {
  const tree = shallow(<OfferList
    offers={[{
      id: 0,
      title: ``,
      type: `room`,
      price: 0,
      img: `room.jpg`,
      rating: 0,
      isPremium: false,
      isFavorite: false,
    }]}
    activeOfferMouseEnterHandler={() => {}}
    deactiveOfferMouseLeaveHandler={() => {}}
    changeOfferFavorite={() => {}}
    getLogin={() => {}}
    classOfferCard="main_offer"
    classCard={{
      list: ``,
      card: ``,
      wrapper: ``,
      info: ``
    }}
  />, createContext());

  expect(toJSON(tree)).toMatchSnapshot();
});
