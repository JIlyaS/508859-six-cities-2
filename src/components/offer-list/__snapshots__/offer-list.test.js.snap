// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfferList correctly renders after relaunch 1`] = `
<div
  className="cities__places-list places__list tabs__content"
>
  <OfferCard
    classCard={
      Object {
        "card": "cities__place-",
        "info": "",
        "list": "cities__places-list places__list tabs__content",
        "wrapper": "cities",
      }
    }
    isFavoriteOffer={false}
    key="0"
    offer={
      Object {
        "id": 0,
        "img": "room.jpg",
        "isFavorite": false,
        "isPremium": false,
        "price": 0,
        "rating": 0,
        "title": "",
        "type": "room",
      }
    }
    offerId={0}
    onActiveOfferMouseEnter={[Function]}
    onDeactiveOfferMouseLeave={[Function]}
    onFavoriteOfferClick={[Function]}
  />
</div>
`;
